### DISCUSSION TITLE
Some problems with API Rules.

### DISCUSSION THREAD

**Discussion by canaldkz** [pocketbase/pocketbase#2814] [Q&A] [answered]
I have several related collections: operations, updates, operations_view, users.
operations_view is a VIEW TABLE. 
I have all kinds of tasks stored in operations, only admins can add and modify them.
In updates are stored different versions of these operations, with the current status, and the performer of the task.
In operations_view, these 2 tables are connected: task name and description are taken from operations, and performer and status are taken from the last created record in updates (by related_to column). Thus, operations_view contains the "latest version of the task".

The database looks like this: 
![screenshot](https://github.com/pocketbase/pocketbase/assets/40259217/203c97c3-b5c9-4abc-8152-9ffb0979129a)

I wanted only those users who are currently the executor of the operation they want to update to can add new records. That is, when adding a record, the executor field nested in the related_to field should be equal to the user's id. 

To implement this, I wrote the following rule for adding records: 
`(@request.auth.verified = true) 
&& (@request.auth.id = @request.data.author.id) 
&& ((@request.auth.role = "editor") 
  || (@request.auth.role = "mod") 
  || (@request.auth.role = "admin")
  || (@request.auth.id = @request.data.view_model.executor))`

But now, the user can add records to updates regardless of whether they are the executor in the latest version or not. The main condition is to write own id in the executor field.
Thanks for any help!

**Comment by ganigeorgiev** [comment]
I'm not sure that I understand the described example and the expected behavior.

For which colletion the above API rule is defined, the `updates`?
What is `@request.data.view_model`?

**Reply by canaldkz** [reply]
Ah, forgot to mention. "view_model" is a relation field to "operation_view" collection. Act same as "related_to" field. And yes, rule defined for "updates".

**Reply by ganigeorgiev** [reply]
How do you maintain the relation relation to the view?
Could you share your view query?

Keep in mind that the Create API rule is executed after a "dry submit" of the record, aka. the record will be already inserted in the database, meaning that it can affect your view and the api rule will only decide whether to commit the transaction or rollback.

**Reply by canaldkz** [reply]
Expected pipeline: 
1) Admin adds new task to "operations" collection, with defined "name" and "description". Nothing else.
2) Admin/Mod/Editor adds new update in "updates" collection, that related to this new task.
In this new update, they defines "state" and "executor" for task. For example, status = "in_progress", executor = %id-of-related-user%.
3) After this actions, in operations_view (dont forget, this is a view collection) we have a new record that contains: "name", "description", "status", "executor". First 2 fields inherits value from record in "operations" collection. Last 2 field inherits value from latest created record in "updates" collection that have same id in "related_to" field. 

If you wonder, there a query for operations_view collection: 

`SELECT op.id, op.name op.description, u.executor, u.status`
`FROM operations op`
`LEFT JOIN (`
`  SELECT related_to, MAX(created) AS max_created`
`  FROM updates`
`  WHERE status IS NOT NULL AND status <> ''`
`  GROUP BY related_to`
`) last_e ON op.id = last_e.related_to`
`LEFT JOIN updates u ON last_e.related_to = u.related_to AND last_e.max_created = u.created;`

**Reply by canaldkz** [reply]
> Keep in mind that the Create API rule is executed after a "dry submit" of the record, aka. the record will be already inserted in the database, meaning that it can affect your view and the api rule will only decide whether to commit the transaction or rollback.

I suspected this for some reason, perhaps you have a solution for this occasion?

**Reply by ganigeorgiev** [reply, marked as answer]
If you want to allow only the last `updates` executor, you may not need 2 `updates` joins, but I admit that I don't fully understand the requirements.

As mentioned earlier, the `updates` view query in your case will be affected by the create operation. There is no direct fix for that as it is an implementation detail how the create rule execute.

Without changing your DB structure, the only workaround I can think of at the moment is to attach to the [Record CRUD hooks](https://pocketbase.io/docs/event-hooks/#record-crud-api-hooks) and perform the access check using the non-transactional `app.Dao()` (with the next v0.17.0 there will be option to bind to the hooks with JS).
